src/devices/cpu/avr8/avr8ops.hxx

```diff
@@ -1115,93 +1181,105 @@ void avr8_device::op_dec(uint16_t op)
 	m_r[RD5(op)] = res;
 }
 
-void avr8_device::op_jmp(uint16_t op)
+template <int NumTimers>
+void avr8_device<NumTimers>::op_jmp(uint16_t op)
 {
 	uint32_t offs = KCONST22(op) << 16;
-	m_pc++;
-	m_shifted_pc += 2;
-	offs |= m_program->read_word(m_shifted_pc);
-	m_pc = offs;
-	m_pc--;
+	m_pc += 2;
+	uint16_t wordval = m_program->read_word(m_pc);
+	offs |= wordval;
+	m_pc = offs << 1;
+	m_pc -= 2;
 }
 
-void avr8_device::op_call(uint16_t op)
+template <int NumTimers>
+void avr8_device<NumTimers>::op_call(uint16_t op)
 {
-	push((m_pc + 2) & 0x00ff);
-	push(((m_pc + 2) >> 8) & 0x00ff);
+	push(((m_pc >> 1) + 2) & 0x00ff);
+	push((((m_pc >> 1) + 2) >> 8) & 0x00ff);
 	uint32_t offs = KCONST22(op) << 16;
-	m_pc++;
-	m_shifted_pc += 2;
-	offs |= m_program->read_word(m_shifted_pc);
-	m_pc = offs;
-	m_pc--;
+	m_pc += 2;
+	offs |= m_program->read_word(m_pc);
+	m_pc = offs << 1;
+	m_pc -= 2;
 }
 
-void avr8_device::op_ret(uint16_t op)
+template <int NumTimers>
+void avr8_device<NumTimers>::op_ret(uint16_t op)
 {
 	m_pc = pop() << 8;
 	m_pc |= pop();
-	m_pc--;
+	m_pc = (m_pc << 1) - 2;
 }
 
-void avr8_device::op_reti(uint16_t op)
+template <int NumTimers>
+void avr8_device<NumTimers>::op_reti(uint16_t op)
 {
 	m_pc = pop() << 8;
 	m_pc |= pop();
-	m_pc--;
+	m_pc = (m_pc << 1) - 2;
 	SREG_W(AVR8_SREG_I, 1);
 }
 
-void avr8_device::op_sleep(uint16_t op)
+template <int NumTimers>
+void avr8_device<NumTimers>::op_sleep(uint16_t op)
 {
-	m_pc--;
+	m_pc = (m_pc << 1) - 2;
 }
```

Comment:
Sorry, I assume I’m missing something here, but why does `op_sleep` need to adjust the program counter shift?  I’m assuming `op_ret` and `op_reti` need to because they’re converting a word address from the stack to an internal word address format, but `op_sleep` is just modifying the value in the program counter.  Why does it need adjustment in this case?

Reply:
Should be addressed now.